python/cuml/utils/input_utils.py

```diff
@@ -248,3 +248,95 @@ def order_to_str(order):
         return 'column'
     elif order == 'C':
         return 'row'
+
+
+def input_to_host_array(X, order='F', deepcopy=False,
+                        check_dtype=False, convert_to_dtype=False,
+                        check_cols=False, check_rows=False,
+                        fail_on_order=False):
+    """
+    Convert input X to host array (NumPy) suitable for C++ methods that accept
+    host arrays.
+    Acceptable input formats:
+    * Numpy array - returns a pointer to the original input
+    * cuDF Dataframe - returns a deep copy always
+    * cuDF Series - returns by reference or a deep copy depending on
+        `deepcopy`
+    * cuda array interface compliant array (like Cupy) - returns a
+        reference unless deepcopy=True
+    * numba device array - returns a reference unless deepcopy=True
+
+    Returns: namedtuple('host_array', 'array pointer n_rows n_cols dtype')
+
+    `host_array` is a new device array if the input was not a NumPy device
+        array. It is a reference to the input X if it was a NumPy host array
+    """
+
+    if convert_to_dtype:
+        X = convert_dtype(X, to_dtype=convert_to_dtype)
+        check_dtype = False
+
+    if isinstance(X, cudf.DataFrame):
+        dtype = np.dtype(X[X.columns[0]]._column.dtype)
+        if order == 'F':
+            X_m = X.as_gpu_matrix(order='F')
+        elif order == 'C':
+            X_m = cuml.utils.numba_utils.row_matrix(X)
```

Comment:
is it faster to do this on GPU? I thought we were pretty slow there...

Reply:
row_matrix is pretty fast actually, the cudf version for order=C not su much...